### PR TITLE
Rewrite References api in ES9

### PR DESCRIPTION
**Pull request by mrkvon**
Starts fixing #804 

So far the new eslint rules are: `no-var`, `prefer-const`, `arrow-spacing`.

### Migrate server code to ES6+
1. in `.eslintrc` add the path of the (migrated file or folder) to `overrides[0].files`
2. run `npx eslint --fix ./path/to/migrated/file_or_folder`
3. fix any pending eslint errors
4. use Promises with (or without) async functions; arrow functions; spread operators; ...; remove dependency on `async` library
5. run linter again as needed

The `ecmaScript` in `.eslintrc` override is currently set to version `9` (`ES2018`). The code is compatible with ES7 or 8.

### CODE REVIEW

**Comment by mrkvon** [comment]
So far about 30% less lines of code and significantly less nesting.

**Comment by mrkvon** [comment]
Something may come up, but generally I consider this done.

It will be mergeable after merging #761. Only if the move to ES7+ is desirable already.

**Comment by nicksellen** [comment]
You can override eslint config via glob matching since v4.1.0, see https://eslint.org/docs/user-guide/configuring. This would mean can just have one config file in the root that can apply to all server stuff.

Also, if you nicked just the babel conversion from https://github.com/Trustroots/trustroots/pull/790 could apply to clientside too :) ... but another PR I think.

**Comment by mrkvon** [comment]
> override eslint config via glob matching

will do :+1:

File: .eslintrc.js [hunk 1]
```diff
@@ -74,5 +74,21 @@ module.exports = {
...
     _: false,
     AppConfig: true
-  }
+  },
+  overrides: [{
+    // ES 2018 - specify migrated files and folders here
+    files: [
+      'testutils/data.server.testutils.js',
+      'modules/references/server/**',
+      'modules/references/tests/server/**'
```
Comment (simison) [inline .eslintrc.js]: Nice and simple 👌

File: .eslintrc.js [hunk 2]
```diff
@@ -74,5 +74,21 @@ module.exports = {
...
+    ],
+    parserOptions: {
+      ecmaVersion: 9
```
[the remark this answers is not in the document]
Comment (mrkvon) [reply .eslintrc.js]: I'd prefer to keep the version here (more strict). Would you agree @simison ?

I'll rebase and fix tomorrow.

[the remark this answers is not in the document]
Comment (simison) [reply .eslintrc.js]: Works for me 👍

File: .eslintrc.js (hunk 1 above)
Comment (mrkvon) [reply .eslintrc.js]: Yes, @nicksellen 's idea. :bulb: I had additional `.eslintrc` all around the place. :)

**Comment by mrkvon** [comment]
ES2018 is now available.
It will be nice if further code is written like this.
Manual coming soon, hopefully today.
:tada:

@simison what do you think about `import` with `.mjs` in nodejs? shall we stick with `require` for a while, or pursue this?

**Comment by mrkvon** [comment]
Ah, I forgot to put a [link to the node documentation](https://nodejs.org/api/esm.html).
I believe that if we want to run ES6 modules in Node, we need to use `--experimental-modules` and filename extension `mjs` instead of `js`. Are you aware of another way?

**Comment by simison** [comment]
Right, thanks for elaborating! Today I learned about these experimental modules. :-)

Sounds _little too exotic_, frankly. We could risk running into issues with tooling that might still be evolving around these methods (production server/linters/editors and IDEs). I'd say let's follow [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principles when possible?

Hopefully those APIs get stable soon because they sound amazing. :-) With Babel, we're on the good track of not getting stuck with a specific version of ES, as we did with ES5.

**Comment by mrkvon** [comment]
> YAGNI

:heart:

> Moving to simple `import ... from`'s is already a good step forward.

I believe [`import ... from` is not possible in node](https://stackoverflow.com/questions/45854169/how-can-i-use-an-es6-import-in-node), unless we use `mjs` + `--experimental-modules`; or babel.
I also believe it's an overkill to use babel for server-side nodejs code; so for me it's either `require` or `import + mjs`. These beliefs may be false.

**Comment by simison** [comment]
Oh yea, you're right — I did not know that. :-) I haven't clearly worked enough with ES6 in the backend.

Looks like we'd have to run `babel-cli` to get non `.mjs` files (I so dislike renaming files). Let's see how far we can get with require — it'll allow us keeping additional layers and off and keep things simple.

**Pull request merged**